native/src/main/java/io/ballerina/lib/ldap/Ldap.java

```diff
@@ -128,6 +189,57 @@ public static Object getEntry(BObject ldapClient, BString distinguishedName, BTy
         }
     }
 
+    public static Object search(Environment env, BObject ldapClient, BString baseDN,
+                                BString filter, BString scope) {
+        try {
+            SearchScope searchScope = getSearchScope(scope);
+            LDAPConnection ldapConnection = (LDAPConnection) ldapClient.getNativeData(NATIVE_CLIENT);
+            validateConnection(ldapConnection);
+            Future future = env.markAsync();
+            SearchResultListener searchResultListener = new CustomSearchResultListener(future, baseDN.getValue());
+            SearchRequest searchRequest = new SearchRequest(searchResultListener, baseDN.getValue(),
+                                                            searchScope, filter.getValue());
+            ldapConnection.asyncSearch(searchRequest);
+            return null;
+        } catch (LDAPException e) {
+            return Utils.createError(e.getMessage(), e);
+        }
+    }
+
+    public static Object searchWithType(Environment env, BObject ldapClient, BString baseDN,
+                                        BString filter, BString scope, BTypedesc typeParam) {
+        try {
+            SearchScope searchScope = getSearchScope(scope);
+            LDAPConnection ldapConnection = (LDAPConnection) ldapClient.getNativeData(NATIVE_CLIENT);
+            validateConnection(ldapConnection);
+            Future future = env.markAsync();
```

Comment:
I assume that `validateConnection` is doing a network call. In that case shall we do like this:
```suggestion
            Future future = env.markAsync();
            validateConnection(ldapConnection);
```

We do not want to block the incoming thread for the validate call IMO. Check for other places as well.